test/Microsoft.AspNet.Mvc.FunctionalTests/PrecompilationTest.cs

```diff
@@ -162,6 +162,97 @@ public class PrecompilationTest
                 File.WriteAllText(Path.Combine(viewsDirectory, "_GlobalImport.cshtml"), globalContent.TrimEnd(' '));
             }
         }
+        
+        [Fact]
+        public async Task PrecompiledView_RendersWithoutSourceFile()
+        {
+            // Arrange
+            IServiceCollection serviceCollection = null;
+            var server = TestHelper.CreateServer(_app, SiteName, services =>
+            {
+                _configureServices(services);
+                serviceCollection = services;
+            });
+            var client = server.CreateClient();
+
+            var serviceProvider = serviceCollection.BuildServiceProvider();
+            var applicationEnvironment = serviceProvider.GetRequiredService<IApplicationEnvironment>();
+
+            var viewsDirectory = Path.Combine(applicationEnvironment.ApplicationBasePath, "Views", "Home");
+            var layoutContent = File.ReadAllText(Path.Combine(viewsDirectory, "Layout.cshtml"));
+            var indexContent = File.ReadAllText(Path.Combine(viewsDirectory, "Index.cshtml"));
+            var viewstartContent = File.ReadAllText(Path.Combine(viewsDirectory, "_ViewStart.cshtml"));
+            var globalContent = File.ReadAllText(Path.Combine(viewsDirectory, "_GlobalImport.cshtml"));
+            
+            // We will render a view that writes the fully qualified name of the Assembly containing the type of
+            // the view. If the view is precompiled, this assembly will be PrecompilationWebsite.
+            var assemblyNamePrefix = GetAssemblyNamePrefix();
+            
+            try
+            {
+                // Act - 1
+                var response = await client.GetAsync("http://localhost/Home/Index");
+                var responseContent = await response.Content.ReadAsStringAsync();
+
+                // Assert - 1
+                Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                var parsedResponse1 = new ParsedResponse(responseContent);
+                Assert.StartsWith(assemblyNamePrefix, parsedResponse1.ViewStart);
+                Assert.StartsWith(assemblyNamePrefix, parsedResponse1.Layout);
+                Assert.StartsWith(assemblyNamePrefix, parsedResponse1.Index);
+
+                // Act - 2
+                // Delete the Layout file and verify it is still served.
+                await DeleteFile(viewsDirectory, "Layout.cshtml");
```

Comment:
This doesn't actually cover the change you made in the code path. The first time the view is served, `IsValidatedPreCompiled` is set to `true` - so we don't use the code path that you made the change in. What I don't follow though is that we're serving views that have been deleted from disk after we added file watchers on it. That sounds like a bug and might be something we need to address.

What your test might need to do would be closer along the lines of

```C#
try
{
   // Act
   await DeleteFile(viewsDirectory, "_ViewStart.cshtml");
   await DeleteFile(viewsDirectory, "_Layout.cshtml");
   await DeleteFile(viewsDirectory, "Index.cshtml");

   var response = await client.GetAsync("http://localhost/Home/Index");

   // Assert
  
    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
    var parsedResponse1 = new ParsedResponse(responseContent);
    Assert.StartsWith(assemblyNamePrefix, parsedResponse1.ViewStart);
    Assert.StartsWith(assemblyNamePrefix, parsedResponse1.Layout);
    Assert.StartsWith(assemblyNamePrefix, parsedResponse1.Index);
```